### PR TITLE
Migrated to latest Dart Version and null safety

### PR DESCRIPTION
**Pull request by escamoteur**
Hi,

while porting WebGL to Flutter I used your repo as a source for samples. While doing that I wanted to see how the original looks like so I updated everything to latest Dart with null safety.

Cheers
Thomas

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -2,7 +2,10 @@ name: learn_gl
...
+environment:
+  sdk: ">=2.12.0-0 <3.0.0"
+
 dev_dependencies:
-  build_runner: ^1.2.8
-  build_web_compilers: ^1.2.0
-  pedantic: ^1.5.0
+  build_runner: ^1.10.0
+  build_web_compilers: ^2.11.0
+  pedantic: ^1.9.0
```
Comment (creativecreatorormaybenot) [inline pubspec.yaml:11]: `1.10.0` is the latests/null safety release :)

```suggestion
  pedantic: ^1.10.0
```

**Comment by jtmcdole** [comment]
Howdy!  Let me take a look!

File: web/json_object.dart [hunk 2]
```diff
@@ -24,21 +24,17 @@ part of learn_gl;
...
 
   JsonObject(String fromJson) {
     Map data = json.decode(fromJson);
 
-    List<dynamic> numArray = data['vertexNormals'];
+    List<dynamic>? numArray = data['vertexNormals'];
     if (numArray != null) {
-      List<double> normals =
-          new List<double>.from(numArray.map((index) => index.toDouble()));
+      List<double> normals = new List<double>.from(numArray.map((index) => index.toDouble()));
```
Comment (jtmcdole) [inline web/json_object.dart:37]: Can you drop the `new`?

File: web/learn_gl.dart [hunk 3]
```diff
@@ -41,35 +41,31 @@ part 'lesson6.dart';
...
+part 'rectangle.dart';
+// Some of our objects that we're going to support
+part 'renderable.dart';
+part 'sphere.dart';
+part 'star.dart';
 
-CanvasElement canvas = querySelector("#lesson01-canvas");
-RenderingContext gl;
-Lesson lesson;
+CanvasElement canvas = querySelector("#lesson01-canvas") as CanvasElement;
```
Comment (jtmcdole) [inline web/learn_gl.dart:53]: Left hand side variable isn't needed if we cast on the RHS

File: web/learn_gl.dart [hunk 4]
```diff
@@ -82,23 +78,22 @@ void main() {
...
   if (urlParameters.containsKey("lsn")) {
     defaultLesson = int.parse(urlParameters["lsn"]);
   }
 
-  SelectElement lessonSelect = querySelector("#lessonNumber");
+  SelectElement lessonSelect = querySelector("#lessonNumber") as SelectElement;
   for (int i = 1; i < 17; i++) {
-    lessonSelect.children.add(new OptionElement(
-        data: "Lesson $i", value: "$i", selected: defaultLesson == i));
+    lessonSelect.children.add(new OptionElement(data: "Lesson $i", value: "$i", selected: defaultLesson == i));
```
Comment (jtmcdole) [inline web/learn_gl.dart:91]: >80 col. I like the flutterism of adding a trailing `,` (flutter didn't exist when I wrote this library)

File: web/lesson1.dart [hunk 5]
```diff
@@ -50,26 +50,20 @@ class Lesson1 extends Lesson {
 
     // bindBuffer() tells the WebGL system the target of future calls
     gl.bindBuffer(WebGL.ARRAY_BUFFER, triangleVertexPositionBuffer);
-    gl.bufferData(
-        WebGL.ARRAY_BUFFER,
-        new Float32List.fromList(
-            [0.0, 1.0, 0.0, -1.0, -1.0, 0.0, 1.0, -1.0, 0.0]),
+    gl.bufferData(WebGL.ARRAY_BUFFER, new Float32List.fromList([0.0, 1.0, 0.0, -1.0, -1.0, 0.0, 1.0, -1.0, 0.0]),
```
Comment (jtmcdole) [inline web/lesson1.dart:53]: did dartfmt not fix these?

File: web/lesson4.dart [hunk 6]
```diff
@@ -18,15 +18,41 @@ part of learn_gl;
...
+  Pyramid pyramid = Pyramid();
+  Cube cube = Cube();
 
-  num rPyramid = 0.0, rCube = 0.0;
+  double rPyramid = 0.0, rCube = 0.0;
 
   Lesson4() {
-    program = new GlProgram(
+    program = GlProgram(
+      // '''
```
Comment (jtmcdole) [inline web/lesson4.dart:30]: intentional comment?

File: web/lesson7.dart [hunk 7]
```diff
@@ -157,11 +154,11 @@ class Lesson7 extends Lesson {
     mvPopMatrix();
   }
 
-  num xSpeed = 3.0, ySpeed = -3.0;
-  num xRot = 0.0, yRot = 0.0;
-  num z = -5.0;
+  double xSpeed = 3.0, ySpeed = -3.0;
```
Comment (jtmcdole) [inline web/lesson7.dart:157]: left-hand-side not needed. var is good.

**Review by jtmcdole — COMMENTED** [review]
Thanks for doing this; a few formatting comments and I think a left over program?

**Comment by escamoteur** [comment]
Sure will make it. check out https://github.com/FlutterGL/flutter_web_gl

File: web/lesson4.dart (hunk 6 above)
Comment (escamoteur) [reply web/lesson4.dart:30]: removed

**Review by jtmcdole — APPROVED** [review]
Solid. Thanks for the link to FlutterGL - I'm excited to see that being added; and super happy its using FFI. The older github.com/google/dart-gl uses ye'olde native plugins.

**Pull request merged**